Replace Dataset struct with Sample and MiniBatch

yeewen1018 opened the pull request:
**Overview:**
This PR refactors the original `Dataset` struct into two modular abstractions: 
* `Sample`: Represents **one** data example (`HashMap<String, Tensor>`)
* `MiniBatch`: Represents a batch of samples, formed by stacking tensors from multiple `Sample`s 

**Problem:**
The previous `Dataset` struct directly stored batched tensors (i.e., mini-batch), rather than modelling a dataset as a collection of individual examples. This complicated future extensions such as streaming datasets or proper batching abstractions. 

**Solution:**
```text
New structure:
src/
├── sample.rs        # `Sample` struct
├── minibatch.rs   # `MiniBatch` struct
├── lib.rs               # module declarations
```
| Type       | Key methods                           |
|------------|---------------------------|
| `Sample`   | `new()`, `from_single()`, `get()`|
| `MiniBatch`| `from_samples()`, `to_device()` |

Validation checks for `MiniBatch::from_samples()`: 
- Rejects empty input 
- Requires matching feature keys across samples  
- Verifies stack-compatible tensor shapes 

Note: Padding logic for variable-length sequences is assumed to be handled externally (i.e., in a future `DataLoader`). 

**Usage Example**

```rust
// Create samples
let samples = vec![
    Sample::from_single("input_ids", tokens_tensor),
    Sample::from_single("input_ids", tokens_tensor2),
];

// Create validated batch
let batch = MiniBatch::from_samples(samples)?;
let gpu_batch = batch.to_device(Device::cuda_if_available());
```

**Tests:**
* Tests are co-located with the corresponding module. 
* `sample.rs`: basic construction with multiple fields
* `minibatch.rs`: batching behaviour and device transfer 

**Future work:**
* Implement `Dataset` trait 
* Implement `DataLoader`
* Evaluate performance for streaming/iterable datasets.

yeewen1018 committed:
feat: Add Dataset struct for regular tensors

nikg7 commented on crates/data_preparation/src/lib.rs:
```diff
@@ -0,0 +1,151 @@
+    /// Validates that:
+    /// - No tensors are scalar (must have at least one dimension).
+    /// - All tensors share the same size for the first dimension (batch size).
+    pub fn new(tensors: HashMap<String, Tensor>) -> Result<Self, String> {
```
use `anyhow::Result<Self>` ?

yeewen1018 replied:
Done! Switched to `anyhow::Result` in [66b61bf](https://github.com/oumi-ai/roumi/commit/66b61bf65c66628f473e227601d385b0c0506f23). Thank you for the suggestion.

nikg7 commented on crates/data_preparation/src/lib.rs:
```diff
@@ -0,0 +1,151 @@
+use tch::Tensor;
+
+#[derive(Debug)]
+pub struct Dataset {
```
Is it really a dataset or just a single example with multiple features? 

Consider adding a comment with documentation on the format e.g., tensor shapes

yeewen1018 replied:
Updated in [e68bfab](https://github.com/oumi-ai/roumi/commit/e68bfab4e44bdb967a79201f015192b26fbd502b) to clarify struct definition and tensor shape rules.

yeewen1018 committed:
refactor: Change Dataset::new to use anyhow::Result

Update the `Dataset::new` constructor to return `anyhow::Result<Self>`
instead of the previous `Result<Self, String>`. This provides a more
ergonomic error handling.

yeewen1018 committed:
docs: update Dataset struct documentation to clarify batch semantics and tensor shape requirements.

nikg7 commented on crates/data_preparation/src/lib.rs:
```diff
@@ -0,0 +1,175 @@
+///    training/inference, this batch can be split into smaller mini-
+///    batches. *(Implementation note: Mini-batch splitting for future work.)*
+#[derive(Debug)]
+pub struct Dataset {
```
A couple of high-level comments:
1. `Dataset` is currently defined as `struct`. Should we introduce Dataset `trait` (~abstract interface) to decouple interface from implementation?
2.  There are two types of datasets:  a. Datasets that can be fully loaded into memory b. Iterable datasets e.g., very large datasets (TB-s) used for pre training , multimodal datasets (dataset with images, video, audio, etc) https://huggingface.co/docs/datasets/en/about_mapstyle_vs_iterable  . We may want to support both types 
3.  Currently, dataset value is `Tensor` .  The tensors can be very large as they contain multidimensional uncompressed data (e.g., `float`-s) . In practice , it may be useful to also support compressed data e.g., PNG bytes for images (much smaller than equivalent `Tensor`)  . WDYT?

yeewen1018 replied:
Thank you for the insightful feedback! After further reading, here is what I’m thinking: 

**Proposed Changes**
We can separate the concerns into three parts: 

1. **Data representation** 
* Rename the current `Dataset` struct as `MiniBatch`
* Introduce a new `Sample` struct to represent individual training examples 

2. **Dataset** 
* Implement it as a **trait**: 
    * `InMemoryDataset`: For small datasets that fit in memory
    * `IterableDataset`: For streaming large datasets 
* Implement `load_dataset(…)` as a builder function to unify both entry points 
```mermaid
classDiagram
    class DatasetBuilder {
        +load_dataset(path: &Path, streaming: bool) -> Result~Box~dyn Dataset~~
    }

    class Dataset {
        <<trait>>
        +iter_samples() -> impl Iterator~Item = Result~Sample~~
    }

    class InMemoryDataset {
        -samples: Vec~Sample~
        +iter_samples() -> impl Iterator~Item = Result~Sample~~
    }

    class IterableDataset {
        -shards: Vec~Path~
        +iter_samples() -> impl Iterator~Item = Result~Sample~~
    }

    DatasetBuilder ..> Dataset : creates
    Dataset <|-- InMemoryDataset
    Dataset <|-- IterableDataset
```
3. **Training interface** 
* Implement a `DataLoader` to handle batching. 
* The idea is to separate data access from batching. 
* `Dataset` manages storage, whereas `DataLoader` handles batching/shuffling/device movement? 

```mermaid
classDiagram
    class DataLoader {
        +new(dataset: Box~dyn Dataset~, batch_size: usize) -> Self
        +iter_batches() -> impl Iterator~Item = Result~MiniBatch~~
    }

    DataLoader o-- Dataset : composes
    DataLoader ..> MiniBatch
```

Would this structure be okay? Or let me know if you would recommend any adjustments. Thanks!

nikg7 replied:
this sounds reasonable to me 

@taenin WDYT ?

taenin replied:
+1, I think this sounds reasonable! Appreciate the diagrams!

yeewen1018 replied:
Okay! Will work on this later this week.

yeewen1018 replied:
I just modified the original `Dataset` into `Sample` and `MiniBatch`, and updated the PR description to reflect the changes. So this PR is ready for your review whenever convenient.

nikg7 commented on crates/data_preparation/src/lib.rs:
```diff
@@ -0,0 +1,175 @@
+use std::collections::HashMap;
+use tch::Tensor;
+
+/// The `Dataset` struct serves as a container for machine learning data,
```
The way it's defined it looks more like "Mini-batch", not a complete "dataset".  Is that how you plan to use it? Should we rename it to that extent ? 

Is batch_size expected to be small, or can be large e.g., millions/billions of records  ?

yeewen1018 committed:
refactor: rename Dataset struct to MiniBatch

yeewen1018 committed:
feat: add Sample struct to represent individual data example

yeewen1018 committed:
refactor!: modify MiniBatch struct to be constructed from Sample and add unit tests

- Implemented `MiniBatch::from_samples(Vec<Sample>)` as the new constructor
- Removed old constructor based on pre-batched tensors
- Include unit tests for construction, device transfer, and error cases

yeewen1018 committed:
refactor: split Sample and MiniBatch into separate modules

- Moved `Sample` from `src/lib.rs` to `src/sample.rs`
- Moved `MiniBatch` from `src/lib.rs` to `src/minibatch.rs`
- Updated `src/lib.rs` to expose both modules
- Left unit tests inline within each module for simplicity

nikg7 commented on crates/data_preparation/src/minibatch.rs:
```diff
@@ -0,0 +1,187 @@
+                .collect();
+
+            // Validate that tensor shapes are compatabile for stacking
+            let reference_shape = tensors_to_stack[0].size();
```
just a warning: it's not always possible to simply `stack()` features as some features may have variable shapes in practice.

There is an abstraction called `collator`  that is used to combine  features from  multiple examples. It can be model-dependent.

Would it be possible to introduce `Collator` as a new interface (`trait`) which takes an array of samples and produced `Minibatch` . 

`stack(...)`-based collaotr may be the most simple default collator.

yeewen1018 replied:
I just added the `Collator` trait with an intial `StackCollator`:
- Moved the stacking logic from `MiniBatch::from_samples` to `StackCollator`
- Replaced `MiniBatch::from_samples(samples)` with `MiniBatch::collate(sample, Collator)`. 

Will implement the padding collator later, but this should give us the basic interface for now. Here is the commit:[13c1cbc](https://github.com/oumi-ai/roumi/commit/13c1cbc924a19e7a8e92fc023a012b9b15c726e4)

Let me know if you would make any adjustments!

yeewen1018 replied:
Hi @nikg4 , I have just pushed the commit for a padding collator ([bf011fa](https://github.com/oumi-ai/roumi/pull/15/commits/bf011faf5e659e5c28358bb716d96ad088704a65)). A few notes on the implementation choice:
- For the current padding strategy, I am pre-allocating each sample’s full padded shape and then using `narrow(dim, start, len).copy_(&source)` to insert the original data. This should be better than chaining `Tensor::cat` calls, which will lead to multiple intermediate allocations and cause peak-memory spikes. 
- A better way of doing this is of course using `slice-assign`, but community feedback showed that it is slow and it is no longer supported by `tch` (at least I can't find it). [Here is the discussion link](https://github.com/huggingface/candle/issues/1441)
- At very large batch sizes, this per-tensor loop may bottleneck. We can later explore other ways to optimize this or introduce a slice-assign variant(?) 
- Also the collator needs the full batch to compute maximum length, so it is not streaming-friendly. I think a workaround is to have a bucket sampler? If so, we can work on this later. 

Let me know if you have any suggestions! Otherwise I think I will park padding optimizations for now and move on to the `Dataset` trait in a new, follow-up PR?

yeewen1018 committed:
feat!: introduce `Collator` trait and `StackCollator`; replace `MiniBatch::from_samples` constructor with `MiniBatch::collate`

yeewen1018 committed:
refactor(stackcollator): simplify feature key validation check

yeewen1018 committed:
feat(collator): add PaddingCollator for variable length data

- Implemented `PaddingCollator` in `src/collator.rs`
- Support three padding modes per feature/dimension:
  - `Dynamic` – pad to the longest example in the batch
  - `Fixed(n)` – pad or truncate to exactly `n`
  - `Symmetric(n)` – center-pad with preference to right/bottom
- Expose a `pad(feature, rules, Optional<pad_value>)` builder
- Add tests for:
  - 1D sequence padding (dynamic/fixed)
  - 3D image padding (fixed/symmetric)
  - Multi-feature alignment (e.g., input_ids + attention_mask)
  - Error cases (shape mismatches, invalid dims)

nikg7 commented on crates/data_preparation/src/collator.rs:
```diff
@@ -0,0 +1,397 @@
+pub struct StackCollator;
+
+impl Collator for StackCollator {
+    fn collate(&self, samples: Vec<Sample>) -> Result<MiniBatch> {
```
`Vec<Sample>`  type transfers ownership to this function , which I think is unnecessary here. Can we use a slice instead ? https://doc.rust-lang.org/std/vec/struct.Vec.html#slicing

nikg7 commented on crates/data_preparation/src/collator.rs:
```diff
@@ -0,0 +1,397 @@
+        let first_keys: HashSet<&String> = samples[0].features.keys().collect();
+        for sample in &samples {
+            if sample.features.keys().collect::<HashSet<_>>() != first_keys {
+                bail!("Mismatched feature keys across samples");
```
would be good to include offending keys into error message

nikg7 commented on crates/data_preparation/src/collator.rs:
```diff
@@ -0,0 +1,397 @@
+        }
+
+        // Validate feature keys
+        let first_keys: HashSet<&String> = samples[0].features.keys().collect();
```
optional (I don't know this enough). Is it really necessary to do `.keys().collect()` , which copies all key? 

Can we use pre-existing `.keys()` to do these validations? (w/o copying)

nikg7 commented on crates/data_preparation/src/collator.rs:
```diff
@@ -0,0 +1,397 @@
+pub enum PaddingRule {
+    Dynamic,        // Pad to maximum size of batch
+    Fixed(i64),     // Pad (or truncate) to exactly this size
+    Symmetric(i64), // Pad evenly to both sides.
```
tbh, I haven't seen `Symmetric` padding used anywhere for such applications. Did you? 

and it's not very well defined  if padding length is odd . Let's say you have to add 3 PAD values, do you add 1 left + 2 right, or 2 left + 1 right?

but I did see `left_side` vs `right_side` padding .  Let's add support for left-side padding

nikg7 commented on crates/data_preparation/src/collator.rs:
```diff
@@ -0,0 +1,397 @@
+/// Defines how a tensor should be padded across a batch
+#[derive(Debug)]
+pub enum PaddingRule {
+    Dynamic,        // Pad to maximum size of batch
```
Let's rename `Dynamic` to `MaxLength` ?

nikg7 commented on crates/data_preparation/src/collator.rs:
```diff
@@ -0,0 +1,397 @@
+    ///     - `PaddingRule::Fixed(n)` = pad to a fixed length `n` (right-side/bottom-side)
+    ///     -`PaddingRule::Symmetric(n)` = pad to a fixed length `n` with equal padding on both sides.
+    /// - `pad_values`: optional padding value (defaults to 0.0)
+    pub fn pad<I>(mut self, feature: impl ToString, rules: I, pad_values: Option<f64>) -> Self
```
why the last arg is called `pad_values` (plural) if it's a single optional value ?

yeewen1018 committed:
refactor(collator): modify `collate` to take `&[Sample]` instead of `Vec<Sample>`

yeewen1018 committed:
feat(collator): add missing/extra keys reporting to key validation check

yeewen1018 committed:
refactor(collator): rename `PaddingRule::Dynamic` to `PaddingRule::MaxLength`

yeewen1018 committed:
feat(collator): add PaddingRule::FixedLeft for left/top padding

- Rename PaddingRule::Fixed -> PaddingRule::FixedRight for explicit semantics
- Update match arms and padding logic to handle both FixedLeft/FixedRight
- Update documentation for all padding rules
- Add test to validate FixedLeft padding behaviour

yeewen1018 committed:
fix(collator): rename `pad_values` param to `pad_value` for singular consistency

yeewen1018 commented:
@nikg4 Thank you for the detailed feedback! I have just addressed all of them in the following commits: 
1. [2f9f9ff](https://github.com/oumi-ai/roumi/pull/15/commits/2f9f9ff1116c91c82fd8469853ec9d737bdca98e) - Changed `Vec<Sample>` to `&[Sample]`
2. [b63b3e0](https://github.com/oumi-ai/roumi/pull/15/commits/b63b3e0787d79d71da2eca4b04d7db97534ef471) - Key validation with error message showing missing and extra found keys
3. [5c9a42d](https://github.com/oumi-ai/roumi/pull/15/commits/5c9a42d1e176174041bcce7b7a92e5790fcb5eb9) - Renamed `PaddingRule::Dynamic` to `PaddingRule::MaxLength`.
4. [15c50bd](https://github.com/oumi-ai/roumi/pull/15/commits/15c50bdad84cf1a9f0c10a31b56fe50356deabd2) - Split `PaddingRule::Fixed` into `PaddingRule::FixedRight` (existing right-pad behaviour) and `PaddingRule::FixedLeft` (left-pad behaviour). 
5. [33a9f04](https://github.com/oumi-ai/roumi/pull/15/commits/33a9f04e3e26ae2b19f2360aac07941469de4b24) - Fixed `pad_values` to `pad_value`. 

A few notes: 
1. Kept the HashSet since a single allocation enables O(1) lookup per key, this is cheaper than a linear search per key without it. 
2. The motivation I had for implementing symmetric padding is to handle image processing use-cases, in which center-padding is used. I think, the convention is to do floor/ceil arithmetic to compute the padding distribution. And so the extra pad usually goes to the right/bottom. So for an odd padding value (e.g., 3), we would pad 1 to the left, and 2 to the right.
